### ISSUE TITLE
ERRORS WHILE CONVERTING MODEL TO TORCHSCRIPT

### ISSUE THREAD

**Issue by bobbilichandu** [WongKinYiu/ScaledYOLOv4#167]
Faced below error while converting the model to torchscript. 

```
Starting TorchScript export with torch 1.7.1...
TorchScript export failure: SyncBatchNorm expected input tensor to be on GPU

```
Following [https://github.com/linghu8812/tensorrt_inference/blob/master/ScaledYOLOv4/export_onnx.py](https://github.com/linghu8812/tensorrt_inference/blob/master/ScaledYOLOv4/export_onnx.py), I edited my code as follows.

```
import argparse

import torch

from utils.google_utils import attempt_download
from experimental import attempt_load
import common

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--weights', type=str, default='./yolov4-p5.pt', help='weights path')
    parser.add_argument('--img-size', nargs='+', type=int, default=[640, 640], help='image size')
    parser.add_argument('--batch-size', type=int, default=1, help='batch size')
    opt = parser.parse_args()
    opt.img_size *= 2 if len(opt.img_size) == 1 else 1  # expand
    print(opt)

    # Input
    img = torch.zeros((opt.batch_size, 3, *opt.img_size))  # image size(1,3,320,192) iDetection

    # Load PyTorch model
    # attempt_download(opt.weights)
    # model = torch.load(opt.weights, map_location=torch.device('cpu'))['model'].float()
    model = attempt_load(opt.weights, map_location=torch.device('cuda'))

    for k, m in model.named_modules():
        m._non_persistent_buffers_set = set()  # pytorch 1.6.0 compatability
        if isinstance(m, common.Conv) and isinstance(m.act, common.Mish):
            m.act = Mish()  # assign activation
        if isinstance(m, common.BottleneckCSP) or isinstance(m, common.BottleneckCSP2) \
                or isinstance(m, common.SPPCSP):
            if isinstance(m.bn, nn.SyncBatchNorm):
                bn = nn.BatchNorm2d(m.bn.num_features, eps=m.bn.eps, momentum=m.bn.momentum)
                bn.training = False
                bn._buffers = m.bn._buffers
                bn._non_persistent_buffers_set = set()
                m.bn = bn
            if isinstance(m.act, common.Mish):
                m.act = Mish()  # assign activation
    model.eval()
    model.model[-1].export = True  # set Detect() layer export=True
    # y = model(img)  # dry run

    # TorchScript export
    try:
        print('\nStarting TorchScript export with torch %s...' % torch.__version__)
        f = opt.weights.replace('.pt', '.torchscript.pt')  # filename
        ts = torch.jit.trace(model, img)
        ts.save(f)
        print('TorchScript export success, saved as %s' % f)
    except Exception as e:
        print('TorchScript export failure: %s' % e)
```
Starting TorchScript export with torch 1.7.1...
TorchScript export failure: SyncBatchNorm expected input tensor to be on GPU

Still facing the same error. Is there a successful way to convert the model to torchscript?

**Comment by sirhikesalot** [comment]
I've overcome this error. After loading in images on the following line
```
 # Input
    img = torch.zeros((opt.batch_size, 3, *opt.img_size))  # image size(1,3,320,192) iDetection
```

We need to move the input tensor to gpu with the following

```
 # Input
    img = torch.zeros((opt.batch_size, 3, *opt.img_size))  # image size(1,3,320,192) iDetection
    img.cuda()

```

@chandu1263 my earlier comment was incorrect. Please accept this change, thx.

**Comment by wolfworld6** [comment]
> Faced below error while converting the model to torchscript.
> 
> ```
> Starting TorchScript export with torch 1.7.1...
> TorchScript export failure: SyncBatchNorm expected input tensor to be on GPU
> ```
> 
> Following https://github.com/linghu8812/tensorrt_inference/blob/master/ScaledYOLOv4/export_onnx.py, I edited my code as follows.
> 
> ```
> import argparse
> 
> import torch
> 
> from utils.google_utils import attempt_download
> from experimental import attempt_load
> import common
> 
> if __name__ == '__main__':
>     parser = argparse.ArgumentParser()
>     parser.add_argument('--weights', type=str, default='./yolov4-p5.pt', help='weights path')
>     parser.add_argument('--img-size', nargs='+', type=int, default=[640, 640], help='image size')
>     parser.add_argument('--batch-size', type=int, default=1, help='batch size')
>     opt = parser.parse_args()
>     opt.img_size *= 2 if len(opt.img_size) == 1 else 1  # expand
>     print(opt)
> 
>     # Input
>     img = torch.zeros((opt.batch_size, 3, *opt.img_size))  # image size(1,3,320,192) iDetection
> 
>     # Load PyTorch model
>     # attempt_download(opt.weights)
>     # model = torch.load(opt.weights, map_location=torch.device('cpu'))['model'].float()
>     model = attempt_load(opt.weights, map_location=torch.device('cuda'))
> 
>     for k, m in model.named_modules():
>         m._non_persistent_buffers_set = set()  # pytorch 1.6.0 compatability
>         if isinstance(m, common.Conv) and isinstance(m.act, common.Mish):
>             m.act = Mish()  # assign activation
>         if isinstance(m, common.BottleneckCSP) or isinstance(m, common.BottleneckCSP2) \
>                 or isinstance(m, common.SPPCSP):
>             if isinstance(m.bn, nn.SyncBatchNorm):
>                 bn = nn.BatchNorm2d(m.bn.num_features, eps=m.bn.eps, momentum=m.bn.momentum)
>                 bn.training = False
>                 bn._buffers = m.bn._buffers
>                 bn._non_persistent_buffers_set = set()
>                 m.bn = bn
>             if isinstance(m.act, common.Mish):
>                 m.act = Mish()  # assign activation
>     model.eval()
>     model.model[-1].export = True  # set Detect() layer export=True
>     # y = model(img)  # dry run
> 
>     # TorchScript export
>     try:
>         print('\nStarting TorchScript export with torch %s...' % torch.__version__)
>         f = opt.weights.replace('.pt', '.torchscript.pt')  # filename
>         ts = torch.jit.trace(model, img)
>         ts.save(f)
>         print('TorchScript export success, saved as %s' % f)
>     except Exception as e:
>         print('TorchScript export failure: %s' % e)
> ```
> 
> Starting TorchScript export with torch 1.7.1...
> TorchScript export failure: SyncBatchNorm expected input tensor to be on GPU
> 
> Still facing the same error. Is there a successful way to convert the model to torchscript?

as for SyncBatchNorm， if you do this bn = nn.BatchNorm2d(m.bn.num_features, eps=m.bn.eps, momentum=m.bn.momentum)，the refrence is same as the before？
